### PR TITLE
[TIMOB-1072] Android: Added RefreshControl support to ListView and TableView.

### PR DESCRIPTION
**Pull request by jquick-axway**
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-1072

**Notes:**

- Does support `RefreshControl.tintColor` property.
- Does not support `RefreshControl.title` property.
- Does not support ScrollView.
  * Modified to log a "not supported" warning when attempting to set `ScrollView.refreshControl` property.
- ListView/TableView `refreshControl` property can be set dynamically like iOS.
  * Setting property to null will disable refresh support.
- Modified TiUISearchBar to support both EditText and TextInputLayout labels.
  * Resolves bug with [TIMOB-24639](https://jira.appcelerator.org/browse/TIMOB-24639)
  * Fix needed in order to test TableView refreshing with SearchBar.

**ListView Test:**

1. Run [TIMOB-1072](https://jira.appcelerator.org/browse/TIMOB-1072) script `RefreshListViewTest.js` on Android.
2. Attempt to swipe down on the ListView.
3. Verify that a refresh control is **_NOT_** displayed.
4. Tap the "Enable Refresh" button.
5. Swipe down on the ListView.
6. Verify that a refresh control is displayed and a new row is added 2 seconds later.
7. Swipe down on the ListView.
8. Verify that a red refresh control is displayed.
9. Swipe down on the ListView.
10. Verify that a green refresh control is displayed.
11. Swipe down on the ListView.
12. Verify that a blue refresh control is displayed.
13. Change "tiapp.xml" property "run-on-main-thread" to the opposite value and re-run steps 1-12.

**TableView Test:**

1. Run [TIMOB-1072](https://jira.appcelerator.org/browse/TIMOB-1072) script `RefreshTableViewTest.js` on Android.
2. Attempt to swipe down on the TableView.
3. Verify that a refresh control is **_NOT_** displayed.
4. Tap the "Enable Refresh" button.
5. Swipe down on the TableView.
6. Verify that a refresh control is displayed and a new row is added 2 seconds later.
7. Swipe down on the TableView.
8. Verify that a red refresh control is displayed.
9. Swipe down on the TableView.
10. Verify that a green refresh control is displayed.
11. Swipe down on the TableView.
12. Verify that a blue refresh control is displayed.
13. Change "tiapp.xml" property "run-on-main-thread" to the opposite value and re-run steps 1-12.

**ScrollView Test:**

1. Run [TIMOB-1072](https://jira.appcelerator.org/browse/TIMOB-1072) script `RefreshVScrollViewTest.js` on Android.
2. Tap on the "Enable Refresh" button.
3. Verify that the following message appears in the Android log...

```
Ti.UI.ScrollView does not support a RefreshControl on Android, yet.
```

### CODE REVIEW

**Review by hansemannn — APPROVED** [review]
My FT **passed** (awaiting QE-test as well). 

A few more comments in the backport-PR!

File: android/modules/ui/src/java/ti/modules/titanium/ui/RefreshControlProxy.java [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+	 * Once assigned, this object's JavaScript APIs can control the view's progress indicator.
+	 * <p>
+	 * The caller must call RefreshControlProxy.unassignFrom() to detach the given view from
+	 * this refresh control once done with it or to disable pull-down refresh support.
+	 * <p>
+	 * If the given view was assigned to another refresh control, then it'll be automatically
+	 * unassigned from it before being assigned to this refresh control.
+	 * <p>
+	 * If this refresh control is currently assigned to another view, then it will be automatically
+	 * unassign from the previous view before being assigned the given view.
```
Comment (ypbnv) [inline android/modules/ui/src/java/ti/modules/titanium/ui/RefreshControlProxy.java]: "unassign" should be "unassigned"

File: android/modules/ui/src/java/ti/modules/titanium/ui/RefreshControlProxy.java [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+ */
+package ti.modules.titanium.ui;
+
+import android.graphics.Color;
+import java.util.HashSet;
+import org.appcelerator.kroll.annotations.Kroll;
+import org.appcelerator.kroll.common.Log;
+import org.appcelerator.kroll.KrollDict;
+import org.appcelerator.kroll.KrollProxy;
+import org.appcelerator.titanium.TiApplication;
```
Comment (ypbnv) [inline android/modules/ui/src/java/ti/modules/titanium/ui/RefreshControlProxy.java]: Unused _org.appcelerator.titanium.TiApplication_ import.

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/TiListView.java [hunk 3]
```diff
@@ -55,7 +58,6 @@
 import android.widget.ListView;
```
Comment (ypbnv) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/TiListView.java]: I have left an unused _FrameLayout_ import from https://github.com/appcelerator/titanium_mobile/pull/9353 before the _ListView_ import. Could you, please, remove it for me?

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableView.java [hunk 4]
```diff
@@ -20,11 +20,13 @@
 import org.appcelerator.titanium.view.TiCompositeLayout;
 import org.appcelerator.titanium.view.TiUIView;
 
+import ti.modules.titanium.ui.RefreshControlProxy;
```
Comment (ypbnv) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableView.java]: Unused import.

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableView.java [hunk 5]
```diff
@@ -42,7 +44,7 @@
 import android.widget.FrameLayout;
```
Comment (ypbnv) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableView.java]: _FrameLayout_ is no longer used.

**Review by ypbnv — APPROVED** [review]
CR: Pass
I have pinned some minor things above.

**Comment by jquick-axway** [comment]
Updated PR based on @hansemannn and @ypbnv 's feedback. Thanks guys!

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIScrollView.java [hunk 6]
```diff
@@ -30,12 +31,16 @@
...
 
+
 public class TiUIScrollView extends TiUIView
 {
 	public static final int TYPE_VERTICAL = 0;
 	public static final int TYPE_HORIZONTAL = 1;
 
 	private static final String TAG = "TiUIScrollView";
+	private static final String REFRESH_CONTROL_NOT_SUPPORTED_MESSAGE =
+			"SrollView does not support a RefreshControl on Android.";
```
Comment (hansemannn) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIScrollView.java]: Typoooo

**Review by hansemannn — CHANGES_REQUESTED** [review]

**Review by hansemannn — APPROVED** [review]

**Comment by lokeshchdhry** [comment]
FR Passed.

`RefreshControl` works as expected with `tableView`, `listView`. On `scrollView` it shows a message `Ti.UI.ScrollView does not support a RefreshControl on Android, yet.`

Studio Ver:     4.9.1.201707200100
SDK Ver:        7.0.0 local build
OS Ver:         10.12.3
Xcode Ver:      Xcode 8.3.3
Appc NPM:       4.2.9
Appc CLI:       6.2.3
Ti CLI Ver:     5.0.14
Alloy Ver:      1.9.13
Node Ver:       6.10.1
Java Ver:       1.8.0_101
Devices:        ⇨ google Nexus 5 --- Android 6.0.1
                ⇨ google Pixel --- Android 7.1.1
Emulator:      Android 4.4.2

**Pull request merged**